src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs

```diff
@@ -58,13 +72,16 @@ public static class EnumerableMappingBuilder
         if (BuildElementMapping(ctx) is not { } elementMapping)
             return null;
 
-        if (ctx.Target.ImplementsGeneric(ctx.Types.Stack, out _))
+        if (ctx.Target.ImplementsGeneric(ctx.Types.StackT, out _))
             return new ForEachAddEnumerableExistingTargetMapping(ctx.Source, ctx.Target, elementMapping, nameof(Stack<object>.Push));
 
-        if (ctx.Target.ImplementsGeneric(ctx.Types.Queue, out _))
+        if (ctx.Target.ImplementsGeneric(ctx.Types.QueueT, out _))
             return new ForEachAddEnumerableExistingTargetMapping(ctx.Source, ctx.Target, elementMapping, nameof(Queue<object>.Enqueue));
 
-        if (ctx.Target.ImplementsGeneric(ctx.Types.ICollection, out _))
+        // create a foreach loop with add calls if source is not an array
+        // and  ICollection.Add(T): void is implemented and not explicit
+        // ensures add is not called and immutable types
+        if (!ctx.Target.IsArrayType() && ctx.Target.HasImplicitInterfaceMethod(ctx.Types.ICollectionT, nameof(ICollection<object>.Add)))
```

Comment:
`HasImplicitInterfaceMethod` uses `FindImplementationForInterfaceMember` to get the matching interface method. For some reason this returns null if the symbol is an array type. 

This could be due to how interfaces work with arrays generics, I dont think `GetMembers` works on array types.
```C#
// if the symbol is an array.
symbol.GetMembers().Count() == 0;
```